core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -377,9 +376,9 @@ class LogSegment(val log: FileRecords,
   def timeWaitedForRoll(now: Long, messageTimestamp: Long) : Long = {
     // Load the timestamp of the first message into memory
     if (rollingBasedTimestamp.isEmpty) {
-      val iter = log.shallowEntries.iterator()
+      val iter = log.batches.iterator()
       if (iter.hasNext)
-        rollingBasedTimestamp = Some(iter.next().record.timestamp)
+        rollingBasedTimestamp = Some(iter.next().maxTimestamp)
```

Comment:
Seems like we need to update the scaladoc of this method. Why are we now using `maxTimestamp` instead of the first timestamp?